internal/common/ingest/ingestion_pipeline.go

```diff
@@ -0,0 +1,211 @@
+type HasPulsarMessageIds interface {
+	GetMessageIDs() []pulsar.MessageID
+}
+
+// InstructionConverter should be implemented by structs that can convert a batch of event sequences into an object
+// suitable for passing to the sink
+type InstructionConverter[T HasPulsarMessageIds] interface {
+	Convert(*EventSequencesWithIds) T
+}
+
+// Sink should be implemented by the struct responsible for putting the data in its final resting place, e.g. a
+// database.
+type Sink[T HasPulsarMessageIds] interface {
+	// Store should persist the sink.  The store is responsible for retrying failed attempts and should only return an error
+	// When it is satisfied that operation cannot be retries.
+	Store(T) error
+}
+
+// EventSequencesWithIds consists of a batch of Event Sequences along with the corresponding Pulsar Message Ids
+type EventSequencesWithIds struct {
+	EventSequences []*armadaevents.EventSequence
+	MessageIds     []pulsar.MessageID
+}
+
+// IngestionPipeline is an application tha reads message from pulsar and inserts them into a sink. The pipeline will
+// handle the following automatically:
+//   - Receiving messages from pulsar
+//   - Splitting messages into batches for efficient processing
+//   - Unmarshalling into event sequences
+//   - Acking processed messages
+//   - Gracefully terminating on a SIGTERM
+//
+// Callers must supply two structs, an InstructionConverter for converting event sequences into something that can be
+// exhausted and a Sink capable of exhausting these objects
+type IngestionPipeline[T HasPulsarMessageIds] struct {
+	pulsarConfig           configuration.PulsarConfig
+	metricsConfig          configuration.MetricsConfig
+	pulsarSubscriptionName string
+	pulsarbatchSize        int
+	pulsarbatchDuration    time.Duration
+	converter              InstructionConverter[T]
+	sink                   Sink[T]
+}
+
+func NewIngestionPipeline[T HasPulsarMessageIds](pulsarConfig configuration.PulsarConfig,
+	pulsarSubscriptionName string,
+	pulsarbatchSize int,
+	pulsarbatchDuration time.Duration,
+	converter InstructionConverter[T],
+	sink Sink[T],
+	metricsConfig configuration.MetricsConfig,
+) *IngestionPipeline[T] {
+	return &IngestionPipeline[T]{
+		pulsarConfig:           pulsarConfig,
+		metricsConfig:          metricsConfig,
+		pulsarSubscriptionName: pulsarSubscriptionName,
+		pulsarbatchSize:        pulsarbatchSize,
+		pulsarbatchDuration:    pulsarbatchDuration,
+		converter:              converter,
+		sink:                   sink,
+	}
+}
+
+// Run will run the ingestion pipeline until a SIGTERM is received
+func (ingester *IngestionPipeline[T]) Run() {
+	// create a context that will end on a sigterm
+	ctx := createContextWithShutdown()
+
+	shutdownMetricServer := common.ServeMetrics(ingester.metricsConfig.Port)
+	defer shutdownMetricServer()
+
+	// Waitgroup that wil fire when the pipeline has been torn down
+	wg := &sync.WaitGroup{}
```

Comment:
I recommend using an errgroup. Run all child services in that errgroup.

Reply:
I haven't used an errgroup because I don't think we need it (yet) but I don't think it will be hard to add one if we need one